### ISSUE TITLE
Address the incompleteness of file table results

### ISSUE THREAD

**Issue by puffyCid** [osquery/osquery#7306]
# Blueprint

The current `file` table currently has a limitation on the number rows it can return.
Depending on the input it will not return an accurate count of files/directories.

For example on my system running the query below to get all files/directories on a system returns 3,736 files/directories
```
osquery> select count(*) from file where path like '/%%';
+----------+
| count(*) |
+----------+
| 3736     |
+----------+
```
But there are actually over 2 millions files in on the system and therefore osquery is returning incomplete/inaccurate data ☹️ 
Based on discussions at #7291, it seems this is a globbing/design decision to prevent symlink looping? Running the above query with verbose flags returns multiple logs with symlink loop detected.

In addition, it looks like the `file` table does not return hidden files/directories?
Running, `select * from file where path like '/Users/puffycid/%'` should return multiple hidden directories and files, but it only returns visible files/directories?

Not returning a full data list (or not throwing a warning message?) about the data could lead to inaccurate results/decisions?

For example, if osquery is installed on a linux server and an organization receives alert about a malicious PHP webshell. If the organization wants to search the full system via a full file listing to see if there are any other suspicious/malicious PHP webshells with osquery they may not get accurate results due to the `file` table not always return an accurate table count? This could lead to potentially missing malicious files ☹️ 
Other tables could probably be used to get the data, but the `file` would be one the most efficient ways to check this IMO?

I think it would be great if this limitation/gap was fixed.
Either through a slight table tweak or fixing the globbing logic.

I was doing some research and I saw that boost had a simple directory recursing function that can recurse through directories (ignores symlinks by default)

Its a simple attempt at a possible solution, I thought it could possibly work. I would be interested in any feedback issues/concerns or other ideas.

I know everyone has different priorities on osquery features/PRs, but IMO I think trying come up with a solution for this limitation (either a code fix, clearer documentation, logging messages) is kind of important? 
Since it affects lots of use cases for osquery: forensic, monitoring, device management, software management, generic information security, etc

Again curious on others thoughts and ideas!
thanks!

**Comment by mike-myers-tob** [comment]
Just throwing in my opinion, that (1) the `file` table documentation should indicate that it does not follow symbolic links; (2) the directory traversing should not bail when it encounters a symbolic link that seems like a bug; (3) it should list hidden files too that seems like another bug. I don't think symbolic links should be followed; if there's a valid use-case for needing to explore symbolic link targets maybe there is some kind of alternative for the user like they could issue follow-up queries based on the rows with a `1` in the `symlink` column.

**Comment by Smjert** [comment]
Beyond what @mike-myers-tob said, I think we really need to try to fix the issue without changing the interface and create a breaking change that causes users to have to chance all their queries against that table.

I believe our issue is that we are almost exclusively relying on the `glob` function, which while it's obviously convenient, it doesn't permit us to have control on the search like we need.

Also, the platformGlob function is used in other places, so I wonder if it carries similar issues on other tables.
Finally that specific function is used on all platforms and Windows has a custom implementation, which I wonder how it compare against this (although I recall there were some complaints of it being slow: https://github.com/osquery/osquery/issues/6679).
But in general, I would always try to write an implementation that can be used on all platforms and all places, and not having to reroll it slightly different everytime.

I'm not saying we have to necessarily fix everything before we can fix the `file` table, but I think this has brought up again several possible issues around our internal APIs that work with files.

**Comment by puffyCid** [comment]
all good points so far. I agree with all of them.  In regards to fixing the issue, if sticking with globbing is the preferred method thats fine. But what about an implementation that preserves the original syntax but adds additional options?  
Ex: add `start_path` and `path_limit` and preserve the original `path` and `directory` requirements? And have the table perform different actions depending on which/number of required column(s) are provided?

Ex: `select * from file where start_path = '/' and path like 'Users/%/Downloads' and path_limit = -1`. Osquery would crawl the whole disk and only return/stat results that contain the globbed path `Users/%/Downloads`

But u would still be able to run a query like `select * from file where path like 'Users/<user>/Downloads/%%'` if you wanted to

Im not against sticking exclusively with the globbing method, but that looks/sounds difficult to fix 🙃 ?

**Comment by directionless** [comment]
When you describe things like `start_path` and `path_limit` that conjures a file walk interface to me. Which seems like it might be useful. I don't have any sense about whether it makes sense to extend the `file` table to support it, vs creating a new table.

**Comment by Lorsh** [comment]
I am wondering where we are on this issue? Are there any plans to improve it in the upcoming releases?

**Comment by directionless** [comment]
> I am wondering where we are on this issue? Are there any plans to improve it in the upcoming releases?

As far as I know, this is not something anyone is working on. The core of the `file` is not designed to walk the entire os. I think there's space to create something, but I'm not sure how that will fit. 

I don't see this as a incompleteness in the file table, but as a feature request for a different sort of interaction.

**Comment by JohnSpeno** [comment]
Was this issue fixed in with https://github.com/osquery/osquery/issues/7291 perhaps?

**Comment by directionless** [comment]
#7291 may have fixed part of this, but I think the deeper issue is that globbing is not a suitable way to traverse a large filesystem. At present osquery does not have a filewalk mechanism. 

I think there's a place for a filewalk implementation, and I'd be interested in seeing contributions to that effect. Though I think there are some questions about the interface, events, and asynchronous nature. 

Meanwhile I'm going to close this as WONTFIX. If someone wants to look at filewalks, I'd love to see it.
